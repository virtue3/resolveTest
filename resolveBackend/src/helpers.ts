
import { fileURLToPath } from "url";
import axios from "axios";
import { default as sqlite3 } from "sqlite3";
import fs from "fs";
import path from "path";

import { getRedisClient } from "./redisClient";

import {
  Entity,
  EntityCategory,
  EntityCategoryAttribute,
  Maybe,
} from "./__generated__/graphql";

sqlite3.verbose();

// import.meta.url es2020 having issues with jest / how project is configured, quick fix.
//   really hosed myself with build targets / project settings here and this is not an easy fix for
let metaUrl;
if (process.env.NODE_ENV === "test") {
  metaUrl = "file://";
} else {
  metaUrl = import.meta.url;
}
export const __filename = fileURLToPath(metaUrl);
export const __dirname = path.dirname(__filename);

export type EntityAtttributeRow = {
  entity_name: string;
  id: number;
  name: string;
  category: string;
  data_type: number;
  data_type_context: string;
  description: string;
  display_name: string;
  flags: number;
  display_precision: number;
  attr_vals: string;
};

export function getRowAttrsToGQLAttrs(
  row: EntityAtttributeRow
): EntityCategoryAttribute {
  return {
    entityName: row.entity_name,
    name: row.name,
    dataType: row.data_type,
    dataTypeContext: row.data_type_context,
    description: row.description ?? "",
    displayName: row.display_name,
    flags: row.flags,
    displayPrecision: row.display_precision,
    attrVals: row.attr_vals,
  };
}

export type EntityWithNameAndAttributes = {
  entityName: string,
  entityId: string,
  categories: { [key: string]: EntityCategoryAttribute[] }
}

// must be called inside db.serialize block!!!
export function fetchEntityRow(db: sqlite3.Database, entityId: string): Promise<EntityWithNameAndAttributes> {
  return new Promise((resolve, reject) => {
    const entJSON: EntityWithNameAndAttributes = { entityName: "", entityId, categories: {} };
    db.each(
      // do as much work in sqlite as we can as it's faster and we've already paid the cost of loading it.
      `SELECT oa.*, GROUP_CONCAT(ov.value) as attr_vals
    FROM _objects_attr oa
    JOIN _objects_eav oe ON oa.id = oe.attribute_id
    LEFT JOIN _objects_val ov ON oe.value_id = ov.id
    WHERE oe.entity_id = ?
    GROUP BY oa.id;
    `, [entityId],
      (err, row: EntityAtttributeRow) => {
        if (err) {
          console.error(`error fetching entityID ${entityId}`);
          return reject(err);
        }

        const category = row.category;

        if (category === "__name__") {
          entJSON.entityName = row.attr_vals;
          // filter all categories starting with "__" (internal)
        } else if (!category.match(/^__/)) {
          // let's immediately conver to GQL data to work with
          const GQLRow = getRowAttrsToGQLAttrs(row);
          if (category in entJSON.categories === false) {
            entJSON.categories[category] = [GQLRow];
          } else {
            entJSON.categories[category].push(GQLRow);
          }
        }

      }, (err, _count) => {
        if (err) {
          console.error(`error in db each for fetching entites ${entityId}`)
          return reject(err);
        }
        return resolve(entJSON);
      }
    );
  });
}

export async function exportDBEntitiesAsJson(db: sqlite3.Database): Promise<{ [key: string]: Entity }> {
  return new Promise((resolve, reject) => {
    db.serialize(() => {
      db.all('SELECT DISTINCT entity_id FROM _objects_eav;', [], (err, entityIds: [{ entity_id: string }]) => {
        if (err) {
          return reject(err);
        }

        // building up the giant json of all the entities to dump into redis.
        const entityFetchPromises: [Promise<EntityWithNameAndAttributes>?] = [];

        // iterate through every entity id and create a fetch rows promise for each entity_id
        entityIds.forEach((entityId: { entity_id: string }) => {
          entityFetchPromises.push(fetchEntityRow(db, entityId.entity_id));
        });

        // if this was ~10x bigger we'd have to probably created a seperate redis db
        //  for each "scene file" that we download and load the entities into it via 
        //  hset on each entityID with the same expiry.  Then just do the database reload when the
        //  data is stale.
        Promise.all(entityFetchPromises).then((values) => {
          // convert keyedByEntityId to GQL format so when we store it we do 0 conversion after fetching
          //  if we are going to pay to JSON encode / decode we should pre-process as much as possible
          const GQLEntityHash: { [key: string]: Entity } = {};
          values.forEach((entity) => {
            if(!entity) {
              return;
            }

            const categories = entity.categories;
            const GQLEntityCategories: Maybe<EntityCategory>[] = [];

            for (const key in categories) {
              GQLEntityCategories.push({
                name: key,
                attributes: entity.categories[key]
              });
            }

            GQLEntityHash[entity.entityId] = {
              id: entity.entityId,
              name: entity.entityName,
              categories: GQLEntityCategories
            }
          })

          resolve(GQLEntityHash);
        }).catch((err) => {
          console.error(`Error fetching entities ${err}`);
          reject(err);
        });
      });
    });
  });
}


export async function checkIfFetchIdInRedis(fetchId: string) {
  const dbCached = await getRedisClient()?.redisClient?.exists(fetchId);
  // if dbCached is > 0 then the key exists in redis.
  if(dbCached) {
    return dbCached > 0
  }
  return false;
}

export async function getFetchIdFromRedis(fetchId: string) {
  const dbFromRedis = JSON.parse(await getRedisClient().redisClient?.get(fetchId) ?? "") as { [key: string]: Entity };
  return dbFromRedis;
}

export async function downloadFileToDisk(url: string, outPath: string): Promise<void> {

  const fileStream = fs.createWriteStream(path.resolve(__dirname, outPath));

  return new Promise((resolve, reject) => {
    axios({
      method: 'get',
      url,
      responseType: 'stream',
    })
      .then((response) => {
        // Pipe the response data to the writable stream
        response.data.pipe(fileStream);

        // When the stream finishes (file is completely written), close the file stream
        response.data.on('end', () => {
          // file is written to disk nothing to return since it's on disk.
          resolve();
        });
        response.data.on('error', (err) => {
          reject(err);
        });
      }).catch((err) => {
        reject(err);
      })
  })
}

export async function deleteFilefromDisk(filePath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.unlink(path.resolve(__dirname, filePath), (err) => {
      if (err) {
        return reject(err);
      }
      resolve();
    });
  })
}

export async function loadDbFromDisk(dbPath: string): Promise<sqlite3.Database> {
  return new Promise((resolve, reject) => {
    const db = new sqlite3.Database(path.resolve(__dirname, dbPath), (err) => {
      if (err) {
        console.error("error loading db from disk", err);
        return reject(err);
      }

      resolve(db);
    });
  })
}