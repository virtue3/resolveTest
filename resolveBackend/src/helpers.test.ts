
import { default as sqlite3 } from "sqlite3";

// mock redis
jest.mock('redis');
jest.mock("./redisClient");

import { expect, describe, jest, test, afterEach, beforeEach} from '@jest/globals';
import { exportDBEntitiesAsJson,  loadDbFromDisk,  checkIfFetchIdInRedis, getFetchIdFromRedis } from "./helpers";


describe('Helpers Test Suite', () => {
  let consoleSpy;

  beforeEach(() => {
    if (typeof consoleSpy === "function") {
      consoleSpy.mockRestore();
    }
  });

  afterEach(() => {
    jest.clearAllMocks();
  });


  describe("DB operations", () => {
    test('Load DB From Disk Works', async () => {
      jest.spyOn(sqlite3, 'Database').mockImplementation((...args) => {
        // test code is not type -super- critical, ts-ignores are ok-ish if not super awful.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        setTimeout(() => args[1]());
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return {} as any;
      });

      let db;

      await expect(
        db = loadDbFromDisk("dummy_path")
      ).resolves.not.toThrowError();

      expect(db).toBeDefined();
    });

    test('Load DB from Disk Throws on load error', async () => {
      // eat console error to have clean test output
      consoleSpy = jest.spyOn(console, "error").mockImplementation(() => { });
      jest.spyOn(sqlite3, 'Database').mockImplementation((...args) => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        setTimeout(() => args[1](new Error("Mock Error")));
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return {} as any;
      });

      try {
        await loadDbFromDisk("dummy_path");
      } catch (err) {
        expect(err.message).toBe("Mock Error")
      }
    });
  });

  describe("Redis Operations", () => {
    test("checkIfFetchIdInRedis works", async () => {
      let checked = await checkIfFetchIdInRedis("TEST_KEY_VALID");
      expect(checked).toBeTruthy();

      checked = await checkIfFetchIdInRedis("TEST_KEY_INVALID");
      expect(checked).toBeFalsy();
    });

    test("check getFetchIdFromRedis works", async () => {
      const json = await getFetchIdFromRedis("TEST_KEY_DUMMY_OBJ_STR");
      expect(json).toEqual({ test: 'dummy' });
    })
  });

  describe("sqlite3 operations", () => {
    const dbAllEntityIds = [{ entity_id: 7001 }];

    const rowDatas = [
      {
        id: 1,
        name: 'name',
        category: '__name__',
        data_type: 20,
        data_type_context: null,
        description: null,
        display_name: null,
        flags: 0,
        display_precision: 0,
        attr_vals: 'Rectangular-Column [286350]'
      }
      , {
        id: 113,
        name: 'Image',
        category: 'Identity Data',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Image',
        flags: 0,
        display_precision: 0,
        attr_vals: ''
      }
      , {
        id: 117,
        name: 'Comments',
        category: 'Identity Data',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Comments',
        flags: 0,
        display_precision: 0,
        attr_vals: ''
      }
      , {
        id: 121,
        name: 'Mark',
        category: 'Identity Data',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Mark',
        flags: 0,
        display_precision: 0,
        attr_vals: ''
      }
      , {
        id: 148,
        name: 'Type Name',
        category: 'Identity Data',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Type Name',
        flags: 8,
        display_precision: 0,
        attr_vals: '400x500-fence'
      }
      , {
        id: 328,
        name: 'Base Level',
        category: 'Constraints',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Base Level',
        flags: 0,
        display_precision: 0,
        attr_vals: 'Ground Floor'
      }
      , {
        id: 329,
        name: 'Base Offset',
        category: 'Constraints',
        data_type: 3,
        data_type_context: 'mm',
        description: null,
        display_name: 'Base Offset',
        flags: 0,
        display_precision: 0,
        attr_vals: '-700.0'
      }
      , {
        id: 344,
        name: 'Column Location Mark',
        category: 'Constraints',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Column Location Mark',
        flags: 8,
        display_precision: 0,
        attr_vals: 'B(225)-1(-14767)'
      }
      , {
        id: 345,
        name: 'Top Level',
        category: 'Constraints',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Top Level',
        flags: 0,
        display_precision: 0,
        attr_vals: 'Ground Floor'
      }
      , {
        id: 346,
        name: 'Top Offset',
        category: 'Constraints',
        data_type: 3,
        data_type_context: 'mm',
        description: null,
        display_name: 'Top Offset',
        flags: 0,
        display_precision: 0,
        attr_vals: '3000'
      }
      , {
        id: 347,
        name: 'Column Style',
        category: 'Constraints',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Column Style',
        flags: 0,
        display_precision: 0,
        attr_vals: 'Vertical'
      }
      , {
        id: 348,
        name: 'Moves With Grids',
        category: 'Constraints',
        data_type: 1,
        data_type_context: '',
        description: null,
        display_name: 'Moves With Grids',
        flags: 0,
        display_precision: 0,
        attr_vals: '1'
      }
      , {
        id: 349,
        name: 'Room Bounding',
        category: 'Constraints',
        data_type: 1,
        data_type_context: '',
        description: null,
        display_name: 'Room Bounding',
        flags: 0,
        display_precision: 0,
        attr_vals: '1'
      }
      , {
        id: 350,
        name: 'Enable Analytical Model',
        category: 'Structural',
        data_type: 1,
        data_type_context: '',
        description: null,
        display_name: 'Enable Analytical Model',
        flags: 0,
        display_precision: 0,
        attr_vals: '1'
      }
      , {
        id: 351,
        name: 'Rebar Cover - Top Face',
        category: 'Structural',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Rebar Cover - Top Face',
        flags: 0,
        display_precision: 0,
        attr_vals: 'Rebar Cover 1 <25 mm>'
      }
      , {
        id: 352,
        name: 'Rebar Cover - Bottom Face',
        category: 'Structural',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Rebar Cover - Bottom Face',
        flags: 0,
        display_precision: 0,
        attr_vals: 'Rebar Cover 1 <25 mm>'
      }
      , {
        id: 353,
        name: 'Rebar Cover - Other Faces',
        category: 'Structural',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Rebar Cover - Other Faces',
        flags: 0,
        display_precision: 0,
        attr_vals: 'Rebar Cover 1 <25 mm>'
      }
      , {
        id: 354,
        name: 'Volume',
        category: 'Dimensions',
        data_type: 3,
        data_type_context: 'm^3',
        description: null,
        display_name: 'Volume',
        flags: 8,
        display_precision: 0,
        attr_vals: '0.74'
      }
      , {
        id: 355,
        name: 'Phase Created',
        category: 'Phasing',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Phase Created',
        flags: 0,
        display_precision: 0,
        attr_vals: 'New Construction'
      }
      , {
        id: 356,
        name: 'Phase Demolished',
        category: 'Phasing',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Phase Demolished',
        flags: 0,
        display_precision: 0,
        attr_vals: 'None'
      }
      , {
        id: 357,
        name: 'Room Name',
        category: 'Other',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Room Name',
        flags: 9,
        display_precision: 0,
        attr_vals: ''
      }
      , {
        id: 358,
        name: 'Room Number',
        category: 'Other',
        data_type: 20,
        data_type_context: '',
        description: null,
        display_name: 'Room Number',
        flags: 9,
        display_precision: 0,
        attr_vals: ''
      }
    ]

    const mockDBWorking = {
      serialize: (cb) => cb(),
      all: (_sql, _params, cb) => {
        setTimeout(() => cb(null, dbAllEntityIds));
      },
      each: async (_sql, entityIds, rowCb, finalCb) => {
        const rowPromises = rowDatas.map(row => {
          return Promise.resolve(() => {
            rowCb(null, row);
          });
        });

        // await each promise then call the each() callback
        for (const p of rowPromises) {
          const func = await p;
          func();
        }

        //call the final callback to resolve the db.each interface
        finalCb();
      }
    }

    test("exportDBEntitiesAsJson works", async () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const jsonBlob = await exportDBEntitiesAsJson(mockDBWorking as any);
      expect(jsonBlob).toMatchSnapshot();
    });
  })
});