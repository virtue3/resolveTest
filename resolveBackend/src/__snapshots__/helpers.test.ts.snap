// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers Test Suite sqlite3 operations exportDBEntitiesAsJson works 1`] = `
{
  "7001": {
    "categories": [
      {
        "attributes": [
          {
            "attrVals": "",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Image",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Image",
          },
          {
            "attrVals": "",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Comments",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Comments",
          },
          {
            "attrVals": "",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Mark",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Mark",
          },
          {
            "attrVals": "400x500-fence",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Type Name",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 8,
            "name": "Type Name",
          },
        ],
        "name": "Identity Data",
      },
      {
        "attributes": [
          {
            "attrVals": "Ground Floor",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Base Level",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Base Level",
          },
          {
            "attrVals": "-700.0",
            "dataType": 3,
            "dataTypeContext": "mm",
            "description": "",
            "displayName": "Base Offset",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Base Offset",
          },
          {
            "attrVals": "B(225)-1(-14767)",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Column Location Mark",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 8,
            "name": "Column Location Mark",
          },
          {
            "attrVals": "Ground Floor",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Top Level",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Top Level",
          },
          {
            "attrVals": "3000",
            "dataType": 3,
            "dataTypeContext": "mm",
            "description": "",
            "displayName": "Top Offset",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Top Offset",
          },
          {
            "attrVals": "Vertical",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Column Style",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Column Style",
          },
          {
            "attrVals": "1",
            "dataType": 1,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Moves With Grids",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Moves With Grids",
          },
          {
            "attrVals": "1",
            "dataType": 1,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Room Bounding",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Room Bounding",
          },
        ],
        "name": "Constraints",
      },
      {
        "attributes": [
          {
            "attrVals": "1",
            "dataType": 1,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Enable Analytical Model",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Enable Analytical Model",
          },
          {
            "attrVals": "Rebar Cover 1 <25 mm>",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Rebar Cover - Top Face",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Rebar Cover - Top Face",
          },
          {
            "attrVals": "Rebar Cover 1 <25 mm>",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Rebar Cover - Bottom Face",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Rebar Cover - Bottom Face",
          },
          {
            "attrVals": "Rebar Cover 1 <25 mm>",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Rebar Cover - Other Faces",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Rebar Cover - Other Faces",
          },
        ],
        "name": "Structural",
      },
      {
        "attributes": [
          {
            "attrVals": "0.74",
            "dataType": 3,
            "dataTypeContext": "m^3",
            "description": "",
            "displayName": "Volume",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 8,
            "name": "Volume",
          },
        ],
        "name": "Dimensions",
      },
      {
        "attributes": [
          {
            "attrVals": "New Construction",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Phase Created",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Phase Created",
          },
          {
            "attrVals": "None",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Phase Demolished",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 0,
            "name": "Phase Demolished",
          },
        ],
        "name": "Phasing",
      },
      {
        "attributes": [
          {
            "attrVals": "",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Room Name",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 9,
            "name": "Room Name",
          },
          {
            "attrVals": "",
            "dataType": 20,
            "dataTypeContext": "",
            "description": "",
            "displayName": "Room Number",
            "displayPrecision": 0,
            "entityName": undefined,
            "flags": 9,
            "name": "Room Number",
          },
        ],
        "name": "Other",
      },
    ],
    "id": 7001,
    "name": "Rectangular-Column [286350]",
  },
}
`;
